src/poetry/core/version/markers.py

```diff
@@ -611,6 +646,55 @@ def intersect(self, other: BaseMarker) -> BaseMarker:
     def union(self, other: BaseMarker) -> BaseMarker:
         return union(self, other)
 
+    def intersect_simplify(self, other: BaseMarker) -> BaseMarker | None:
+        """
+        Finds a couple of easy simplifications for intersection on MarkerUnions:
+
+            - intersection with any marker that appears as part of the union is just
+              that marker
+
+            - intersection between two markerunions where one is contained by the other
+              is just the smaller of the two
+
+            - intersection between two markerunions where there are some common markers
+              and the intersection of unique markers is not a single marker
+        """
+        if other in self._markers:
+            return other
+
+        if isinstance(other, MarkerUnion):
+            our_markers = set(self.markers)
+            their_markers = set(other.markers)
+
+            if our_markers.issubset(their_markers):
+                return self
+
+            if their_markers.issubset(our_markers):
+                return other
+
+            if not our_markers.intersection(their_markers):
+                return None
+
+            # use lists instead of sets for deterministic order
```

Comment:
there's some awkwardness with the lists and sets here.  I suspect none of it matters but there are opportunities for algorithmic improvement.

- you're calculating the set intersection, but then later testing membership against the `common_markers` list which is O(n) instead of O(1)
- "use lists instead of sets for deterministic order": but we only care about `unique_markers` and `their_markers` if they collapse on intersection - in which case their order can't possibly matter

So I think this can be simplified to something like
```python
            shared_markers = our_markers.intersection(their_markers)
            if not shared_markers:
                return None

            unique_markers = our_markers - their_markers
            other_unique_markers = their_markers - our_markers
            unique_intersection = MarkerUnion(*unique_markers).intersect(
                MarkerUnion(*other_unique_markers)
            )
            if isinstance(unique_intersection, (SingleMarker, EmptyMarker)):
                # Use list instead of set for deterministic order.
                common_markers = [
                    marker for marker in self.markers if marker in shared_markers
                ]
                return unique_intersection.union(MarkerUnion(*common_markers))
```

(and much the same in `union_simplify()`)